irksome/stepper.py

```diff
@@ -265,3 +297,253 @@ def advance(self):
 
     def solver_stats(self):
         return (self.num_steps, self.num_nonlinear_iterations, self.num_linear_iterations)
+
+
+class AdaptiveTimeStepper(StageDerivativeTimeStepper):
+    """Front-end class for advancing a time-dependent PDE via an adaptive
+    Runge-Kutta method.
+
+    :arg F: A :class:`ufl.Form` instance describing the semi-discrete problem
+            F(t, u; v) == 0, where `u` is the unknown
+            :class:`firedrake.Function and `v` is the
+            :class:firedrake.TestFunction`.
+    :arg butcher_tableau: A :class:`ButcherTableau` instance giving
+            the Runge-Kutta method to be used for time marching.
+    :arg t: A :class:`firedrake.Constant` instance that always
+            contains the time value at the beginning of a time step
+    :arg dt: A :class:`firedrake.Constant` containing the size of the
+            current time step.  The user may adjust this value between
+            time steps; however, note that the adaptive time step
+            controls may adjust this before the step is taken.
+    :arg u0: A :class:`firedrake.Function` containing the current
+            state of the problem to be solved.
+    :arg tol: The temporal truncation error tolerance
+    :arg dtmin: Minimal acceptable time step.  An exception is raised
+            if the step size drops below this threshhold.
+    :arg dtmax: Maximal acceptable time step, imposed as a hard cap;
+            this can be adjusted externally once the time-stepper is
+            instantiated, by modifying `stepper.dt_max`
+    :arg KI: Integration gain for step-size controller.  Should be less
```

Comment:
`Integration gain` may not be a familiar term.  Quick description of it?

Reply:
On the one hand, I'm totally in agreement here.  On the other hand, this doesn't seem like the best place for a tutorial on [PID controllers](https://en.wikipedia.org/wiki/Proportional–integral–derivative_controller).  Suggestions for how to navigate this?  In essence, the step-size controller that we're using is from the literature [here](https://link.springer.com/article/10.1007/BF01934091).  These are the two knobs to twiddle, who default to recommended values from that paper, but I'm not sure how to convey insight into their choices in a doctoring...

Reply:
I took another shot at this, with a bit more info and a reference.  Thoughts?

Reply:
Sure.  I think this is fine.